### DISCUSSION TITLE
Typescript `setState` accepts a partial state, even if the `replace` option is enabled

### DISCUSSION THREAD

**Discussion by Yonom** [pmndrs/zustand#2578] [Bug report] [answered]
The `setState` function of `Store` accepts a `Partial<T>` type, even if the `replace` option is enabled. This has caused me to accidentally ship bugs in production in the past. My suggestion would be to update the `setState` types to conditionally require a non-partial `T` if the second argument of the function is set to true.

Example:

```ts
type MyStore = { foo: string; bar: string };

const useMyStore = create<MyStore>()(() => ({ foo: "hello", bar: "world" }));

useMyStore.setState({}, true); // <- this does not produce an error, even though the new value is invalid
```

Suggested change in `vanilla.ts`

```patch
 type SetStateInternal<T> = {
   _(
     partial: T | Partial<T> | { _(state: T): T | Partial<T> }['_'],
-    replace?: boolean | undefined,
+    replace?: false | undefined,
   ): void
+  _(
+    partial: T | { _(state: T): T }['_'],
+    replace: true,
+  ): void
 }['_']
```

**Comment by dai-shi** [comment, marked as answer]
I agree it's a reasonable fix. Would you like to open a PR please?
I don't know if it causes some troubles in some edge cases.

**Reply by Yonom** [reply]
Of course, will open a PR :)

I adopted it in my (small) repository and it caught a bug I had missed.
So, this breaks existing (incorrect) code; perhaps best to only ship it with the next major release

**Reply by dai-shi** [reply]
Then, send the PR to `v5` branch please.

**Reply by dai-shi** [reply]
#2580
